spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/bigquant/Linear.scala

```diff
@@ -0,0 +1,218 @@
+      init()
+    }
+  }
+
+  def checkAndInit(): Unit = {
+    if (!_init && weight.getNativeStorage == 0L) {
+      init()
+    }
+  }
+
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    require(input.dim() == 1 || input.dim() == 2,
+      "bigquant.Linear: " + ErrorInfo.constrainInputAsVectorOrBatch)
+
+    checkAndInit()
+
+    val batchSize = if (input.dim() == 1) {
+      output.resize(Array(outputSize)) // TODO
+      1
+    } else {
+      output.resize(Array(input.size(1), outputSize))
+      require(inputSize == input.size(2), s"dimension error")
+      input.size(1)
+    }
+
+    data.init(LinearDataParams(batchSize, inputSize), LinearData)
+
+    ev.getType() match {
+      case FloatType => // TODO
+        val src = input.storage().array().asInstanceOf[Array[Float]]
+        val offset = input.storageOffset() - 1
+
+        BigQuant.FCDataInit(data.getNativeStorage, src, offset, batchSize, inputSize,
+          QuantParams.THRESHOLD, BigQuant.NCHW)
+
+        val outputArray = output.storage().array().asInstanceOf[Array[Float]]
+        val outputOffset = output.storageOffset() - 1
+        val weightSumArray = weight.sumOfRow.asInstanceOf[Array[Float]]
+        val weightSumOffset = 0
+        val biasArray = bias.storage().array().asInstanceOf[Array[Float]]
+        val biasOffset = bias.storageOffset() - 1
+
+        BigQuant.InternalMixPrecisionConvolutionGEMM(
+          BigQuant.NCHW, weight.getNativeStorage, data.getNativeStorage, outputArray,
+          outputOffset, weightSumArray, weightSumOffset, biasArray, biasOffset,
+          batchSize, outputSize, 1, 1,
+          QuantParams.FAULT_TOLERANCE)
+
+      case _ => throw new UnsupportedOperationException(s"Only support Float for quantized model")
+    }
+
+    output
+  }
+
+  override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    throw new UnsupportedOperationException(s"Doesn't updateGradInput for quantized model")
+  }
+
+  override def parameters(): (Array[Tensor[T]], Array[Tensor[T]]) = {
+    (Array(weight, bias), Array(empty, empty))
+  }
+
+  override def getParametersTable(): Table = {
+    T(getName() -> T("weight" -> weight, "bias" -> bias))
+  }
+
+  override def equals(obj: Any): Boolean = {
+    if (!super.equals(obj)) {
+      return false
+    }
+
+    if (!obj.isInstanceOf[Linear[T]]) {
+      return false
+    }
+    val other = obj.asInstanceOf[Linear[T]]
+    if (this.eq(other)) {
+      return true
+    }
+
+    weight == other.weight &&
+      bias == other.bias
+  }
+
+  override def hashCode(): Int = {
+    val seed = 37
+    var hash = super.hashCode()
+    hash = hash * seed + weight.hashCode()
+    hash = hash * seed + bias.hashCode()
+
+    hash
+  }
+
+  override def clearState() : this.type = {
+    super.clearState()
+    this
+  }
+
+  override def toString(): String = {
+    s"bigquant.${getPrintName()}($inputSize -> $outputSize)"
+  }
+
+  def release(): Unit = {
```

Comment:
Where is this used?

Reply:
It's only used in test cases now. To release the memory allocated in native code.